### ISSUE TITLE
[15.0] web_editor lose image quality and size optimization

### ISSUE THREAD

**Issue by zeroheure** [odoo/odoo#61619]
Optimized images from graphical team can't be kept.

Impacted versions: 14

Steps to reproduce: Optimize image file size for the web, using Gimp "Save For The Web" plugin or [Mozilla mozjpeg tool](https://github.com/mozilla/mozjpeg) or an optimization service like [ImageOptim free web interface](https://imageoptim.com/online). Upload this image in Odoo.

Current behavior: Odoo can't keep the optimized file size. It will allways increase a lot (for example from 45 Kb to 122 Kb like the optimized attached). Furthermore the quality is lost.

Expected behavior: Be allowed to disable Odoo image optimization. 

See a test image attached.
![carte cadeau recto-R](https://user-images.githubusercontent.com/24126/98855784-dd1a1f80-245c-11eb-95f2-f30f6bdf598b.jpg)

**Comment by fizcris** [comment]
Is there a way to disable automatic image optimization for the website?

**Comment by zeroheure** [comment]
@fizcris not in Odoo14

**Comment by fizcris** [comment]
Sorry for the inconvenience again.  

Is there an optimum pixel size  and dPi for images to avoid optimization ( or at least minimise it). Let's say what is the optimum size for the web logo? As the trial and error process to achieve "ok" quality it is a bit exhausting.

**Comment by Yenthe666** [comment]
Yep, also been bitten by this just now. Not being able to turn it off or reset sucks big time.

**Comment by zeroheure** [comment]
Same behaviour with Odoo 15 (tested on current master). It is even worse as images uploaded through product's form are not resizable.

**Comment by zeroheure** [comment]
Can someone confirm that image "optimization" is done in odoo/addons/web_editor/static/src/js/editor/image_processing.js ?

**Comment by zeroheure** [comment]
@pedrobaeza could you add the 15 and master labels ?

**Comment by zeroheure** [comment]
Dear @seb-odoo purpose of #31208 was (quoting):
> it is important for SEO and for usability in general that the images are as light as possible in size

Indeed, we are able to achieve near WEBP optimizations on JPEG for years (see tools below), while Pillow is not able to handle all optimizations (see Pyllow #539 (https://github.com/python-pillow/Pillow/issues/539)) and destroy our images.
IMHO, advanced users should be able to upload already optimized images and leaving thew untouched. 

For example there is some easy tools that does a much much better job than Odoo/Pillow:

- Gimp's "Save for the web" option and a similar one in Photoshop 
- Free software [ImageOptim](https://imageoptim.com/) and [ImageOptim online](https://imageoptim.com/online), which are based on mozjpeg
- [Mozjpeg](https://github.com/mozilla/mozjpeg) which can gives near webp results with jpeg optimizations)
- etc.

If the `disable_optimization` option button confuse users, a configuration option could be added to allow the display of it. I agree that most users don't understand image optimization. But it is not Odoo job to decide if optimization is Odoo's job or graphic artist one. Specially when all of these can be done in batch then uploaded.

**Comment by seb-odoo** [comment]
Hello,

I think it would be reasonable to add a config param to be able to disable auto-optimize for advanced users.

But I'm no longer working on this part of the code, I think you should see with @odoo/rd-website and/or web editor teams.

**Comment by rdeodoo** [comment]
We generally try to focus on keeping everything as simple as possible, meaning the less options the better.
This is to deliver a better product to the lambda end user that would need a website builder or a small ecommerce, we are talking about something like a hairdresser or a butcher having ~0 knowledge in IT and with computer in general.

But your point is correct, having a way to disable that, so (very) professional users could bypass that behavior would not be a lot of effort on our part, AFAIK.

Note that it would definitely be a hidden option through an ICP. We really don't want to noise any screen with that option, neither in the right panel neither in the backend website setting. Also we don't want people to discover that in the software.
I guess that could be part of an advanced documentation for the website builder at some point.

An ICP which would simply bypass all the image optimizations with a warning in the right panel next to those options hinting that it got disabled at the database level.

I will get back to you about that.

**Comment by ikus060** [comment]
Any news about this feature ? Our designer already provide optimised images exported by Adobe XD. No need to re-optimise them the quality get really bad.

**Comment by rdeodoo** [comment]
> Any news about this feature ? 

Yes. Thanks for the reminder.

I just created a task internally to handle that, someone on the team should start working on it anytime soon.
It will be something similar to what I suggested in my previous message.
We will let you know once it is done.

**Comment by jwaes** [comment]
Any ETA yet ?
Downsampling really affects the image quality. Need to decide to start working on a workaround or not.

**Comment by rdeodoo** [comment]
This one is being worked on right now: https://github.com/odoo/odoo/pull/90619. It is currently waiting for https://github.com/odoo/odoo/pull/89223 to be merged (hopefully tonight or tomorrow) to be merged to finish the task as it will require to introduce some warning in the right panel.

**Comment by Yenthe666** [comment]
@rdeodoo both are going into master so won't benefit the V15 projects though?

**Comment by rdeodoo** [comment]
We prefer to do it in master first in the cleanest way possible and see from there about a potential backport.

**Comment by zeroheure** [comment]
There is a missing point here : in any case, the image "optimized" in Odoo should never be heavier than the original. A simple test on file size is enough. Behaviour would be
- same or greater file size => keep the original uploaded file
- lower file size => overwrite original with Odoo optimized file

**Comment by tranngocson1996** [comment]
@zeroheure Do you have direction to do this part at 15.0? I also want to customize it but have no idea how to code. Thank

**Comment by zeroheure** [comment]
No, not really. 
But an image width of 1024 pixel prevent Odoo to apply "optimization". It was the behaviour  on Odoo 8 to 13. I didn't tested on more recent version.

**Comment by tranngocson1996** [comment]
> There is a missing point here : in any case, the image "optimized" in Odoo should never be heavier than the original. A test on file size is enough. Behaviour would be:
> 
> * same or greater file size => keep the original uploaded file
> * lower file size => overwrite original with Odoo optimized file
> 
> This is simple and should allways be applied, no need for an advanced option.

@zeroheure Thank for reply, but I mean ask about this part

**Comment by rdeodoo** [comment]
> But an image width of 1024 pixel prevent Odoo to apply "optimization". It was the behaviour on Odoo 8 to 13. 

I am not aware of such a thing (not saying it was or wasn't a thing), but that would be really dubious and not something we want at all.
We want all and every single image to be optimized for the end user, as not doing it and leaving too big images would be a big SEO issue. Uploading a 1024 image in a 250 width column is useless, the system will resize it to something like ~500px (it will assume the 250px col could be larger in lower resolution where cols won't be next to each other but on top of each other taking full width). I don't know exactly the rules without checking but that's about it.

> There is a missing point here : in any case, the image "optimized" in Odoo should never be heavier than the original. 

Sure, but that's another issue.
What has been mainly discussed and the need that came up in this thread is for people to be able to disable the odoo automatic optimization (regardless if the "heavier than the original" issue is fixed or not).

Correct me if I am wrong but everyone here is asking about a way to turn off the optimization because already having images optimized (done by someone manually, or with another software etc).
That's also the requested "feature" from the issue report: `Expected behavior: Be allowed to disable Odoo image optimization.`

**Comment by zeroheure** [comment]
>>  There is a missing point here : in any case, the image "optimized" in Odoo should never be heavier than the original.

> Sure, but that's another issue.

Sure (again), but I think it is part of the issue because what everyone need is basicaly a lower file size. Even, a test on file size would save a big part of the issue because there will be no optimization from Odoo. Thus, the disable optimization option will be unneeded in most cases.

**Comment by zeroheure** [comment]
@tranngocson1996 see https://github.com/odoo/odoo/blob/master/odoo/tools/image.py

**Comment by tranngocson1996** [comment]
> @tranngocson1996 see https://github.com/odoo/odoo/blob/master/odoo/tools/image.py

@zeroheure thank, I looked at this file but it can only get the width and height but not the weight of the image to compare

**Comment by zeroheure** [comment]
@tranngocson1996 I think that weight is checked in the web_editor module. Commits mentionned above relate to that module https://github.com/odoo/odoo/pull/90619

**Comment by tranngocson1996** [comment]
@zeroheure Thank you for your enthusiasm 👍🏿

**Comment by zeroheure** [comment]
@tranngocson1996 if you are able to add that simple behaviour to previous Odoo versions, you will save a lot of people headache.

**Comment by tranngocson1996** [comment]
@zeroheure I'm also having a headache with it these days 😞

**Comment by trinhanhngoc** [comment]
@tranngocson1996 Me too.

**Comment by tranngocson1996** [comment]
@trinhanhngoc long time no see bro 😭

**Comment by trinhanhngoc** [comment]
> @trinhanhngoc long time no see bro 😭

I just discovered that uploading an image with width > 1920 will result in poor quality resized images. Tip: avoid uploading images with width > 1920.

**Comment by tranngocson1996** [comment]
@trinhanhngoc you fixed this bug ?

**Comment by tranngocson1996** [comment]
@trinhanhngoc thank you for your solution 😄

**Comment by rdeodoo** [comment]
Hello, we just merged https://github.com/odoo/odoo/commit/c512f098a7a3bb0773e3aede3c71765ffa7bfad2 and https://github.com/odoo/odoo/commit/567e5b58d544e4e56c3a68d148e862a516d77c4f in master (Odoo 16.4)

It's probably not the version you hoped for, but it's what we have decided.
Hopefully, this will be resolving your issue: an image modified source won't be used and the original source will be kept if there is no real modification (size, crop, effect etc) but the modified source would be greater than the original source.

**Comment by alvaro2203** [comment]
Hello, im using odoo v16 and im having the same problem. I want to keep the quality of the images but Odoo limits it. How can I solve it?

**Comment by rdeodoo** [comment]
You could upgrade to Odoo 17.0 to solve it, as we decided to fix it only up from that version for safety reason as we did not want to risk introducing a bug in stable versions at the time.

If you really need to make it work in Odoo 16, you could maybe try some workaround (not tested) like creating a new `ir.attachment` of type URL and uploading your image directly in that attachment (it's possible for the attachment to have both an URL and some data). That should not resize your image AFAIK.
Then you could edit your page HTML to use that attachment source as your image source.
That's obviously technical, not convenient and not sure it's working.

I'd advice to consider upgrading, as Odoo 16 is getting old and is currently our oldest supported version, though.

**Comment by zeroheure** [comment]
@alvaro2203 if you are using the community version, try with [OCB](https://github.com/OCA/OCB/), which is Odoo Community version with some backports from Odoo Community Association. 
Else, the workaround from Romain is a good clever one.
